src/mlpack/methods/ann/vanilla_rbm_impl.hpp

```diff
@@ -0,0 +1,201 @@
+    typename HiddenLayerType>
+RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::RBM(
+    arma::mat& predictors,
+    InitializationRuleType initializeRule,
+    VisibleLayerType visible,
+    HiddenLayerType hidden):
+    visible(visible),
+    hidden(hidden),
+    initializeRule(initializeRule),
+    reset(false)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::Reset()
+{
+  // Since we are dealing with a static stucture no need for visitor
+  // Reset the parameters of all the layers
+  size_t weight = 0;
+  weight+= visible.Parameters().n_elem;
+  parameter.set_size(weight, 1);
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
+  visible.Parameters() = arma::mat(parameter.memptr(), weight, 1, false, false);
+  hidden.Parameters() = arma::mat(parameter.memptr(), weight, 1, false, false);
+
+  visible.Reset();
+  hidden.Reset();
+  reset = true;
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+template<template<typename> class Optimizer>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Train(const arma::mat& predictors, Optimizer<NetworkType>& optimizer)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+
+  if (!reset)
+  {
+    Reset();
+  }
+
+  // Train the model.
+  Timer::Start("rbm_optimization");
+  optimizer.Optimize(parameter);
+  Timer::Stop("rbm_optimization");
+}
+
+ /* 
+  * This function calculates
+  * the free energy of the model
+  */
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    FreeEnergy(const arma::mat&& input)
+{
+  arma::mat output;
+  visible.Forward(std::move(input), std::move(output));
+  SoftplusFunction::Fn(output, output);
+  return  -arma::dot(input, visible.Bias()) - arma::accu(output);
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Evaluate(const arma::mat& /* parameters*/, arma::vec& orderFunction)
+{
+  double freeEnergy, freeEnergyCorrupted;
+  arma::Col<double> output(orderFunction.n_elem);
+  size_t corruptIdx = math::RandInt(0, visible.Bias().n_elem);
+  for (auto i : orderFunction)
+  {
+    // Do not use unsafe col predictor as we change the input
+    arma::vec currentInput = predictors.col(i);
+    freeEnergy = FreeEnergy(std::move(currentInput));
+    currentInput(corruptIdx) =  1 - predictors(corruptIdx);
+    freeEnergyCorrupted = FreeEnergy(std::move(currentInput));
+    output(i) = SoftplusFunction::Fn(freeEnergyCorrupted - freeEnergy + 1);
```

Comment:
``orderFunction`` holds the visitation order of the current batch and is of size batchSize right? So let's say ``orderFunction`` is ``[120, 121, 122, ...]`` I guess I run into an issue if I do ``output(i)``with ``i = 120`` since ``output`` has the same size as ``orderFunction`` which is ``batchSize``. Maybe I missed something?